### DISCUSSION TITLE
Server discovery

### DISCUSSION THREAD

**Discussion by pditommaso** [It4innovations/hyperqueue#417] [General]
Hello team,

Being the first post in this discussion group, let me congratulate you about this project. It really looks promising and starting with a lean and fresh approach to distributed computing. 

I was wondering how the HQ workers discover the server instance? Could not find much in the docs, but my assumption is that it's done via the shared file `$HOME/.hq-server/hq-current/access.json`. Is that correct? 

Have you considered other approaches for environments in which a shared file system is not possible (mostly cloud)?

**Comment by spirali** [comment]
You are right, ``access.json`` is used in the current version. We ourselves do not have a use case for another approach; however, we are open to add an additional mechanism. If you have any particular mechanism in mind, we can start to discuss it.

**Comment by Kobzol** [comment]
Hello :)

Workers discover the server instance using a [server directory](https://it4innovations.github.io/hyperqueue/stable/deployment/server/#server-directory). It can either use a default location, or you can specify it explicitly (you can also run multiple HQ servers concurrently with custom server directories).

This does not necessarily mean that a shared filesystem is required, you can just copy the server directory after the server is started to another node where the worker should run.

Of course we realize that this solution is less than ideal if you don't have a shared filesystem, and we are prepared to add new ways of providing workers/clients with connection information. We were just waiting for the first use-case to see what will our users actually want (we already had a cloud use-case, but for that it was enough to just copy the directory so far).

Both clients and workers require several pieces of data to connect to the server, at the very least hostname and port, but also things like encryption keys.

Let us know what other ways of passing this information you have in mind. We could e.g. pass some kind of "connection string" to the commands.

**Reply by pditommaso** [reply]
Thanks for the reply. Sorry I was missing the server directory docs. 

> This does not necessarily mean that a shared filesystem is required, you can just copy the server directory after the server is started to another node where the worker should run.

This sounds nice. 

> Let us know what other ways of passing this information you have in mind. We could e.g. pass some kind of "connection string" to the commands.

We don't have yet a concrete use case, however the connection string looks an interesting approach. Also it could be useful to use a cloud storage e.g. AWS S3 to share the connection info. 


I'll comment more, once I've collected a more detailed need.

**Comment by Kobzol** [comment]
You can examine https://it4innovations.github.io/hyperqueue/stable/deployment/cloud/ to see how you can connect HQ components without a shared filesystem. Let us know if you have any further questions!
